### PR TITLE
[GStreamer][MSE] Push STREAM_COLLECTION event after flushing

### PR DESCRIPTION
**Pull request by eocanha**
#### 4c9c583641452ecb3807a9fa58103629431710cf
<pre>
[GStreamer][MSE] Push STREAM_COLLECTION event after flushing
<a href="https://bugs.webkit.org/show_bug.cgi?id=262193">https://bugs.webkit.org/show_bug.cgi?id=262193</a>

Reviewed by Alicia Boya Garcia &lt;aboya@igalia.com&gt;.

Flushing the playback pipeline when very first buffers are being pushed downstream causes
STREAM_COLLECTION event to be dropped before reaching decodebin3. As a result playback never
starts after flush (buffer push return not-linked). This happens on initial seek sometimes,
when it is triggered at the same time that first buffers push.

STREAM_COLLECTION is dropped by GstBaseParse element specificaly that is caching events and
sending them with first frame. Unfortunately flushing will drop all pending events if first
frame is not ready yet.

Original author: Andrzej Surdej (<a href="https://github.com/asurdej-comcast)">https://github.com/asurdej-comcast)</a>

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1197">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1197</a>

* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(webKitMediaSrcStreamFlush): Reset hasPushedStreamCollectionEvent at the end of the flush to force sending of the STREAM_COLLECTION event after that. Only for GStreamer versions that don&apos;t have a proper fix yet.

Canonical link: <a href="https://commits.webkit.org/268582@main">https://commits.webkit.org/268582@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/439ed0286484cd0da6f732bc81320a300b397d78

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20098 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20537 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21155 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21997 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18762 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20328 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23779 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20695 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20318 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/20235 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17463 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22846 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17405 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18260 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24514 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18482 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/18436 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22504 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19038 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18234 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4819 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22578 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18863 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp [hunk 1]
```diff
@@ -649,6 +649,19 @@ static void webKitMediaSrcStreamFlush(Stream* stream, bool isSeekingFlush)
             GST_DEBUG_OBJECT(stream->pad.get(), "StreamingMembers mutex taken, using it to set isFlushing = false.");
             streamingMembers->isFlushing = false;
             streamingMembers->doesNeedSegmentEvent = true;
+
```
Comment (philn) [inline Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:663]: Nit: can you remove this extra empty line?

**Review by ntrrgc — APPROVED** [review]
Seems harmless enough. r+

File: Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp [hunk 2]
```diff
@@ -649,6 +649,19 @@ static void webKitMediaSrcStreamFlush(Stream* stream, bool isSeekingFlush)
...
+            if (!webkitGstCheckVersion(1, 22, 0)) {
+                // In older GST versions STREAM_COLLECTION event is delivered to decodebin3
+                // from parsebin src pad probe. On the way, this event is cached inside
+                // parser element (GstBaseParse) and pushed downstream with first frame.
+                // Flushing before first frame is handled by the parser,
+                // the event is dropped from GstBaseParse and never reaches decodebin3.
+                // GST 1.21.3 added STREAM_COLLECTION event handling on decodebin3 sink pad directly
+                // so this workaround is not needed anymore.
+                GST_DEBUG_OBJECT(stream->pad.get(), "Reset hasPushedStreamCollectionEvent");
+                streamingMembers->hasPushedStreamCollectionEvent = false;
```
Comment (ntrrgc) [inline Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:673]: What happens if you let this change occur on current WebKit?

Comment (eocanha) [reply Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:673]: This block is already being executed. WebKit is already running on GStreamer 1.22.

**Pull request merged**